core/src/main/java/org/apache/calcite/rex/RexExecutor.java

```diff
@@ -32,6 +34,22 @@
    * @param reducedValues List to which reduced expressions are appended
    */
   void reduce(RexBuilder rexBuilder, List<RexNode> constExps, List<RexNode> reducedValues);
+
+  /**
+   * Checks if condition first implies (&rArr;) condition second.
+   *
+   * <p>This reduces to SAT problem which is NP-Complete.
+   * When this method says first implies second then it is definitely true.
+   * But it cannot prove that first does not imply second.
+   *
+   * @param rexBuilder Rex builder
+   * @param rowType row type
+   * @param first first condition
+   * @param second second condition
+   * @return true if it can prove first &rArr; second; otherwise false i.e.,
+   * it doesn't know if implication holds
+   */
+  boolean implies(RexBuilder rexBuilder, RelDataType rowType, RexNode first, RexNode second);
```

Comment:
I am not in favor of adding this method in the RexExecutor interface. Many similar methods exist in RexImplicationChecker which is already public API.